File: commands/psychopass.ts
import { SlashCommandBuilder, ChatInputCommandInteraction } from "discord.js";
import { startSpan } from "@sentry/node";
import { PsychoPasses } from "../clients/backend/psychopass/psychoPasses.js";
import { CommunityPsychoPasses } from "../clients/backend/psychopass/communityPsychoPasses.js";
import embedPsychoPass from "../embeds/psychoPass.js";
import embedCommunityPsychoPass from "../embeds/communityPsychoPass.js";

const data = new SlashCommandBuilder()
    .setName("psychopass")
    .setDescription("Get the Psycho-Pass of a User or Server")
    .addUserOption((option) =>
        option.setName("user").setDescription("Gets this User's Psycho-Pass")
    );

async function execute(interaction: ChatInputCommandInteraction) {
    await interaction.deferReply();
    const user = interaction.options.getUser("user");
    if (user) {
        console.log(
            `${interaction.user.username} (${interaction.user.id}) has requested the Psycho-Pass of @${user.username} (${user.id})`
        );
        const psychoPass = await PsychoPasses.read(user.id);
        if (psychoPass === undefined)
            interaction.editReply(
                "The targeted User has not sent a message yet, and does not have a Psycho-Pass"
            );
        else
            interaction.editReply({
                embeds: [
                    embedPsychoPass(
                        psychoPass,
                        interaction.client,
                        interaction.user,
                        user
                    ),
                ],
            });
    } else {
        console.log(
            `${interaction.user.username} (${interaction.user.id}) has requested the Psycho-Pass of Server: ${interaction.guild?.name} (${interaction.guildId})`
        );
        const psychoPass = await CommunityPsychoPasses.read(
            interaction.guildId!
        );
        interaction.editReply({
            embeds: [
                embedCommunityPsychoPass(
                    psychoPass!,
                    interaction.client,
                    interaction.user,
                    interaction.guild!
                ),
            ],
        });
    }
}

async function onExecute(interaction: ChatInputCommandInteraction) {
    startSpan(
        {
            name: "/psychopass",
        },
        () => execute(interaction)
    );
}

export default { data, execute: onExecute };

File: commands/psycho_pass.py
from loguru import logger as log
from discord import (
    ApplicationContext,
    Bot,
    Cog,
    slash_command,
    User,
    SlashCommandOptionType,
    option,
)

from clients.backend.psycho_pass.community_psycho_passes import CommmunityPsychoPasses
from clients.backend.psycho_pass.psycho_passes import PsychoPasses
from embeds.psycho_pass import embed_community_psycho_pass, embed_psycho_pass


class PsychoPass(Cog):

    def __init__(self, bot: Bot) -> None:
        self.bot = bot

    @slash_command(description="get the psycho-pass of a user or server")
    @option(
        "user",
        type=SlashCommandOptionType.user,
        description="gets this user's psycho-pass",
        required=False,
    )
    async def psychopass(self, ctx: ApplicationContext, user: User) -> None:
        log.debug("{} /psychopass user: @{}", ctx.guild_id, user)
        await ctx.defer()
        if user is not None:
            log.info(
                "@{} ({}) has requested the psycho-pass of @{} ({})",
                ctx.user.name,
                ctx.user.id,
                user.name,
                user.id,
            )
            psycho_pass = PsychoPasses.read(user.id)
            await ctx.edit(embed=embed_psycho_pass(psycho_pass, ctx.user, user))
        else:
            log.info(
                "@{} ({}) has requested the area stress level of {} ({})",
                ctx.user.name,
                ctx.user.id,
                ctx.guild.name,
                ctx.guild_id,
            )
            psycho_pass = CommmunityPsychoPasses.read(ctx.guild_id)
            await ctx.edit(
                embed=embed_community_psycho_pass(psycho_pass, ctx.user, ctx.guild)
            )
